component/src/main/java/org/wso2/extension/siddhi/execution/geo/GetCityResolverFunction.java

```diff
@@ -0,0 +1,216 @@
+                        name = "ip",
+                        description = "The IP address that the user need to get the relevant city",
+                        type = {DataType.STRING})
+        },
+        returnAttributes = @ReturnAttribute(
+                description = "The city which is related to the given IP",
+                type = {DataType.STRING}),
+        examples = @Example(
+                description = "This will return the corresponding city to the given ip address",
+                syntax = "define stream IpStream(ip string);\n" +
+                        "from IpStream\n" +
+                        "select geo:getCity(ip) as city\n" +
+                        "insert into outputStream;")
+)
+public class GetCityResolverFunction extends FunctionExecutor {
+    private static final Log log = LogFactory.getLog(GetCityResolverFunction.class);
+
+    private static final String CACHE_SIZE_KEY = "cacheSize";
+    private static final String IS_CACHE_ENABLED = "isCacheEnabled";
+    private static final int DEFAULT_CACHE_SIZE = 10000;
+    private static final String DEFAULT_GEOLOCATION_RESOLVER_CLASSNAME =
+            "org.wso2.extension.siddhi.execution.geo.internal.impl.DefaultDBBasedGeoLocationResolver";
+
+    private GeoLocationResolver geoLocationResolverImpl;
+    private static LRUCache<String, String> cityLRUCache;
+    private static boolean isCacheEnabled = true;
+    private static int cacheSize = 10000;
+    private static boolean isExtensionConfigInitialized;
+
+    /**
+     * The initialization method for {@link FunctionExecutor}, which will be called before other methods and validate
+     * the all configuration and getting the initial values.
+     *
+     * @param attributeExpressionExecutors are the executors of each attributes in the Function
+     * @param configReader                 this hold the {@link FunctionExecutor} extensions configuration reader.
+     * @param siddhiAppContext             Siddhi app runtime context
+     */
+    @Override
+    protected void init(ExpressionExecutor[] attributeExpressionExecutors, ConfigReader configReader,
+                        SiddhiAppContext siddhiAppContext) {
+        if (attributeExpressionExecutors.length != 1) {
+            throw new SiddhiAppValidationException("Invalid no of arguments passed to geoIp:getCity() function, " +
+                    "required 1, but found " + attributeExpressionExecutors.length);
+        }
+
+        Attribute.Type attributeType = attributeExpressionExecutors[0].getReturnType();
+        if (attributeType != Attribute.Type.STRING) {
+            throw new SiddhiAppValidationException("Invalid parameter type found for first argument ip of " +
+                    "geoIp:getCity() function, required " + Attribute.Type.STRING + ", but found " + attributeType
+                    .toString());
+        }
+
+        if (!isExtensionConfigInitialized) {
+            initializeExtensionConfigs(configReader);
```

Comment:
Move the initialization to start () method. and do connection cleaning if any in the stop() method (you can overwrite these methods.)
Don't do connection related things in the init.

Reply:
As per the offline discussion we had, I will keep the implementation as it is. Because we do not have start/stop method in function extension